cpp/daal/src/algorithms/dtrees/forest/oneapi/df_tree_level_build_helper_oneapi.i

```diff
@@ -479,25 +494,42 @@ services::Status TreeLevelBuildHelperOneAPI<algorithmFPType>::doLevelPartition(c
 
     auto & context = services::internal::getDefaultContext();
 
-    auto & kernel = kernelDoLevelPartition;
+    auto & kernel = kernelDoLevelPartitionByGroups;
 
     {
+        DAAL_ASSERT(nNodes <= _int32max);
         DAAL_ASSERT(nFeatures <= _int32max);
 
-        KernelArguments args(5, status);
+        // nNodes * _partitionMaxBlocksNum is used inside kernel
+        DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION(int32_t, nNodes, _partitionMaxBlocksNum);
+
+        // nodeAuxList is auxilliary buffer for synchronization of left and right boundaries of blocks (nElemsToLeft, nElemsToRight)
+        // processed by subgroups in the same node
+        // no mul overflow check is required due to there is already buffer of size nNodes * _nNodeProps
+        DAAL_ASSERT(_auxNodeBufferProps <= _nNodeProps);
+
+        auto nodeAuxList = context.allocate(TypeIds::id<int>(), nNodes * _auxNodeBufferProps, status);
         DAAL_CHECK_STATUS_VAR(status);
+        context.fill(nodeAuxList, 0, status);
+        DAAL_CHECK_STATUS_VAR(status);
+
+        KernelArguments args(7, status);
+        DAAL_CHECK_STATUS_VAR(status);
+
         args.set(0, data, AccessModeIds::read);
         args.set(1, nodeList, AccessModeIds::read);
-        args.set(2, treeOrder, AccessModeIds::read);
-        args.set(3, treeOrderBuf, AccessModeIds::write);
-        args.set(4, static_cast<int32_t>(nFeatures));
+        args.set(2, nodeAuxList, AccessModeIds::readwrite);
+        args.set(3, treeOrder, AccessModeIds::read);
+        args.set(4, treeOrderBuf, AccessModeIds::write);
```

Comment:
Is buffer size checked for those two buffers?

Reply:
nodeAuxList is allocated here.
Assertion for buffer size is also in this function on lines 492, 493